cpp/notify_users.cc

```diff
@@ -45,62 +50,308 @@ void Usage() {
 }
 
 
-class SearchObject {
-    std::vector<std::string> search_terms_;
-    std::vector<std::string> filters_;
-    std::string type_;
-    std::string class_;
-public:
-    explicit SearchObject(const PHPUtil::Object &minSO);
+namespace {
+
+
+enum class ParseState { ARRAY_EXPECTED, OPEN_PAREN_EXPECTED, PARAM_OR_CLOSE_PAREN_EXPECTED, PARAM_OPEN_PAREN_EXPECTED,
+	                PARAM_VALUE_EXPECTED, PARAM_CLOSE_PAREN_EXPECTED };
+
+
+} // unnamed namespace
+
+
+/** Parses structures like the following:
+
+Array
+(
+    [qf] => Array
+        (
+            [0] => title_short^750 title_full_unstemmed^600 title_full^400 title^500 title_alt^200 title_new^100 series^50 series2^30 author^300 author_fuller^150 contents^10 topic_unstemmed^550 topic^500 geographic^300 genre^300 allfields_unstemmed^10 fulltext_unstemmed^10 allfields fulltext description isbn issn
+        )
+
+    [qt] => Array
+        (
+            [0] => dismax
+        )
+
+    [q] => Array
+        (
+            [0] => brxx
+        )
+)
+
+This is a pretty-printed PHP array of arrays data type representing a query.
+*/
+bool ExtractQueryParams(const std::string &php_query_array,
+			std::map<std::string, std::string> * const params_to_values_map)
+{
+    std::vector<std::string> lines;
+    if (StringUtil::SplitThenTrim(php_query_array, '\n', "\t ", &lines) <= 1)
+	return false;
+
+    static RegexMatcher * const param_name_matcher(RegexMatcher::RegexMatcherFactory("\\[([[:lower:]]+)\\]"));
+    ParseState parse_state(ParseState::ARRAY_EXPECTED);
+    std::string last_parem_name;
+    for (std::string line : lines) {
+	StringUtil::Trim(&line);
+	if (line.empty())
+	    continue;
+
+	switch (parse_state) {
+	case ParseState::ARRAY_EXPECTED:
+	    if (line != "Array")
+		return false;
```

Comment:
Wouldn't it be nice to get some error messages if something fails? So you know where and why it fails?